### PR TITLE
feat!: increase share size to 512 bytes

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/825

This is ready for review but we may not merge it b/c we haven't made a final decision on what the share size should be

### CODE REVIEW

File: x/payment/types/wirepayfordata_test.go
Comment (rootulp) [inline x/payment/types/wirepayfordata_test.go:95]: I'm not sure why this test fails now. Attempts to fix it result in a different error: `ErrInvalidShareCommitments` and not `ErrInvalidShareCommit`.

I don't _think_ this test is still necessary because we test for `ErrInvalidShareCommit` [here](https://github.com/celestiaorg/celestia-app/blob/main/x/payment/types/wirepayfordata_test.go#L82-L84) but curious for reviewers thoughts.

**Review by SweeXordious — COMMENTED** [review]
Awesome stuff :rocket: 
Left some questions.

File: pkg/da/data_availability_header_test.go
Comment (rootulp) [reply pkg/da/data_availability_header_test.go:68]: Good suggestion. I just parallelized this test but the difference in time to run is non-noticeable. Since the added `tt := tt` and `t.Parallel()` don't appear to significantly increase test speed, will opt to not parallelize.

Without `t.Parallel()`
```
$ go test -run TestNewDataAvailabilityHeader -count=1
PASS
ok  	github.com/celestiaorg/celestia-app/pkg/da	1.739s
$ go test -run TestNewDataAvailabilityHeader -count=2
PASS
ok  	github.com/celestiaorg/celestia-app/pkg/da	3.031s
$ go test -run TestNewDataAvailabilityHeader -count=3
PASS
ok  	github.com/celestiaorg/celestia-app/pkg/da	4.545s
``` 

With `t.Parallel()`
```
$ go test -run TestNewDataAvailabilityHeader -count=1
PASS
ok  	github.com/celestiaorg/celestia-app/pkg/da	1.797s
$ go test -run TestNewDataAvailabilityHeader -count=2
PASS
ok  	github.com/celestiaorg/celestia-app/pkg/da	3.057s
$ go test -run TestNewDataAvailabilityHeader -count=3
PASS
ok  	github.com/celestiaorg/celestia-app/pkg/da	4.412s
```

File: pkg/shares/reserved_bytes.go
Comment (rootulp) [reply pkg/shares/reserved_bytes.go:21]: > I don't think this case will ever occur, according to docs: https://pkg.go.dev/encoding/binary#PutUvarint

Agreed! Thanks for pointing out. I'll remove this conditional.

> A better check might be bytesWritten == appconsts.CompactShareReservedBytes, but I don't know if the uint64 is small, whether it will write all bytes (even the zero ones) or not.

Since we initialize the reservedBytes slice to be of length `appconsts.CompactShareReservedBytes`, even if the PutUvarint only writes a value to the first byte, the second byte will remain zero. Note this is confirmed by this [test case](https://github.com/rootulp/celestia-app/blob/8ea9c826692c675bca6d67a93a547513a2d36676/pkg/shares/reserved_bytes_test.go#L58)

> Another check might be <= 0 to check if something was written or not. But needs a bit more investigations.

Since a uint64 can't be `< 0`, I don't think this check is necessary.

File: pkg/shares/reserved_bytes.go
Comment (rootulp) [reply pkg/shares/reserved_bytes.go:15]: One does not need to use naked returns when using named return values. Additionally a named return value is by default initialized to its zero value so these lines are unnecessary if they are specified as named return values:

```golang
        reservedBytes := []byte{}
        var err error
```

Agreed this is a cosmetic thing and I think @evan-forbes has similar opinions against named return values so I think we should have an offline discussion or issue to discuss it. My opinion is that:
1. named return values are good because they provide additional documentation in the function signature which is especially helpful when a function returns multiple values. They also enable the author to remove trivial initialization statements (like the two lines above).
2. naked returns are bad because they force the reader to return to the function signature to learn what values are being returned

So I would prefer if we used named return values but avoided naked returns.

Ref:
- https://yourbasic.org/golang/named-return-values-parameters/
- https://go.dev/tour/basics/7

File: pkg/shares/share_splitting_test.go
Comment (rootulp) [reply pkg/shares/share_splitting_test.go:134]: This is just for tests. Some of the shares in this test file are padded with message data (i.e. `0xc`) but I realize now that `pad` is strongly associated with adding trailing `0`s so I'll clarify this with an additional helper.

Comment (rootulp) [reply pkg/shares/share_splitting_test.go]: Resolved in https://github.com/celestiaorg/celestia-app/pull/850/commits/b2ce5421a00c955a402ee3cbc08a416cfa7564cc

File: pkg/shares/reserved_bytes.go
Comment (rootulp) [reply pkg/shares/reserved_bytes.go]: Resolve in https://github.com/celestiaorg/celestia-app/pull/850/commits/b2ce5421a00c955a402ee3cbc08a416cfa7564cc

File: pkg/shares/reserved_bytes.go
Comment (rootulp) [reply pkg/shares/reserved_bytes.go:15]: On second thought, we've already had a discussion on this: https://github.com/celestiaorg/celestia-core/issues/831 @SweeXordious feel free to comment or re-open that issue

File: pkg/da/data_availability_header_test.go
Comment (SweeXordious) [reply pkg/da/data_availability_header_test.go:68]: Awesome :rocket:

File: pkg/shares/reserved_bytes.go
Comment (SweeXordious) [reply pkg/shares/reserved_bytes.go:15]: Aaah, I see. So, the decision is to leave it to developer preference? 
I personally have nothing against any.

Also, considering the initialization in the beginning, then this function can be written as:

```golang
func NewReservedBytes(byteIndex uint64) (reservedBytes []byte, err error) {
	if byteIndex >= appconsts.ShareSize {
		err = fmt.Errorf("byte index %d must be less than share size %d", byteIndex, appconsts.ShareSize)
                return
	}
	reservedBytes = make([]byte, appconsts.CompactShareReservedBytes)
	bytesWritten := binary.PutUvarint(reservedBytes, byteIndex)
	if bytesWritten > appconsts.CompactShareReservedBytes {
		err = fmt.Errorf("byte index %d must occupy less than %d bytes when written as varint", byteIndex, appconsts.CompactShareReservedBytes)
                return
	}
	return
}
```

My point is, if we're using named returns, then we should use them also inside the function, such as:

```
err = fmt.Errorf("byte index %d must occupy less than %d bytes when written as varint", byteIndex, appconsts.CompactShareReservedBytes)
```
instead of:
```
return []byte{}, fmt.Errorf("byte index %d must occupy less than %d bytes when written as varint", byteIndex, appconsts.CompactShareReservedBytes)
```

But I don't have a strong opinion really.

File: pkg/shares/share_splitting_test.go
Comment (SweeXordious) [reply pkg/shares/share_splitting_test.go]: thanks

File: pkg/shares/reserved_bytes.go
Comment (rootulp) [reply pkg/shares/reserved_bytes.go:15]: I think explicit returns are more readable than naked returns so my preference is:

```golang
return []byte{}, fmt.Errorf("byte index %d must occupy less than %d bytes when written as varint", byteIndex, appconsts.CompactShareReservedBytes)
```
 instead of:

```golang
err = fmt.Errorf("byte index %d must be less than share size %d", byteIndex, appconsts.ShareSize)
return
```

but I'll remove usages of named return values from this PR to resolve this

Comment (rootulp) [reply pkg/shares/reserved_bytes.go]: Resolved in https://github.com/celestiaorg/celestia-app/pull/850/commits/8a6a66bb81294774b915b9763983e75a6559f76c

Comment (SweeXordious) [reply pkg/shares/reserved_bytes.go]: Btw, I'm okey with both, I'm just trying to see which one is better. It's alright to keep any from my perspective.
Btw, as a general rule, when I'm implementing stuff, should I use the naked or explicit returns? or it's left as a developer preference?
Thanks a lot for  your help and time :pray:

Comment (rootulp) [reply pkg/shares/reserved_bytes.go]: No worries. Unfortunately it doesn't seem like there's an opinionated Go convention on it so it's left to developer presence. 

Since this codebase largely doesn't use names return params, I think it makes sense to conform to that.

Comment (SweeXordious) [reply pkg/shares/reserved_bytes.go]: Awesome, thanks a lot again :pray:

**Review by SweeXordious — COMMENTED** [review]
Approved. Thanks a lot for discussion :fire: 
Will defer to Evan to approve it as I'm still exploring this part of the codebase.

**Comment by evan-forbes** [comment]
will try to review this soon, and can help debug why `TestFuzzPrepareProcessProposal` is failing. Likely just a silly oversight in the non-interactive defaults code.

**Comment by evan-forbes** [comment]
this is dooooope, can we keep this as a draft until we plan to merge? I will review as if its not a draft tho, and test w/ celestia-node

File: x/payment/types/wirepayfordata_test.go
Comment (evan-forbes) [reply x/payment/types/wirepayfordata_test.go:95]: I think this is just because the after we increase the share size, it makes the commitment to the message the same for different sizes since it can now fit on one row. If we change the test data to create valid pfd with 4000 byte msg instead of 2000, this test passes

**Review by evan-forbes — COMMENTED** [review]
would prefer to try to fix the test as described in the comment, but tbh, I think its fine as well. 

this LGTM, I think we're ready to merge when we need to. Have we tried this out with celestia-node yet? perhaps we could just upgrade arabica to v0.8.0 since its basically ready

**Review by evan-forbes — APPROVED** [review]
8MB blocks, nbd

**Pull request merged**
